### PR TITLE
Development Deployment Manager

### PR DESCRIPTION
**Pull request by lciolecki**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -1000,6 +1002,23 @@ lazy val liteEmbeddedDeploymentManager = (project in lite("embeddedDeploymentMan
       liteKafkaComponents % "test", liteRequestResponseComponents % "test", componentsUtils % "test",
       testUtils % "test", kafkaTestUtils % "test")
 
+lazy val developmentTestsDeploymentManager = (project in development("deploymentManager")).
+  configs(IntegrationTest).
```
Comment (mproch) [inline build.sbt:1006]: I don't think you need this?

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+  override def close(): Unit = {}
+
+  private def changeState(name: ProcessName, stateStatus: StateStatus): Unit =
+    memory.get(name).foreach(processState => {
+      val newProcessState = processState.withStateStatus(stateStatus)
+      memory.update(name, newProcessState)
+      logger.debug(s"Changed scenario $name state from ${processState.status.name} to ${stateStatus.name}.")
+    })
+
+  private def asyncChangeState(name: ProcessName, stateStatus: StateStatus): Unit =
```
Comment (mproch) [inline engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:102]: Maybe actorSystem.scheduler.scheduleOnce() would be nicer? (ActorSystem is provided by DeploymentManagerProvider)

File: build.sbt (hunk 1 above)
Comment (lciolecki) [reply build.sbt:1006]: Removed ;)

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala (hunk 2 above)
Comment (lciolecki) [reply engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:103]: Good point - done.

**Review by mproch — APPROVED** [review]

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+class DevelopmentDeploymentManager(actorSystem: ActorSystem) extends DeploymentManager with LazyLogging {
+
+  private val MinSleepTime = 5
+  private val MaxSleepTime = 12
+
+  private val memory: TrieMap[ProcessName, ProcessState] = TrieMap[ProcessName, ProcessState]()
+  private val random: ThreadLocalRandom = ThreadLocalRandom.current()
+
+  implicit private class ProcessStateExpandable(processState: ProcessState) {
+    def withStateStatus(stateStatus: StateStatus): ProcessState = {
```
Comment (arkadius) [inline engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:34]: We already have `ProcessState.withStatusDetails`. Can't it be used here? (just asking)

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
+
+class DevelopmentDeploymentManager(actorSystem: ActorSystem) extends DeploymentManager with LazyLogging {
+
+  private val MinSleepTime = 5
+  private val MaxSleepTime = 12
+
+  private val memory: TrieMap[ProcessName, ProcessState] = TrieMap[ProcessName, ProcessState]()
+  private val random: ThreadLocalRandom = ThreadLocalRandom.current()
```
Comment (arkadius) [inline engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:31]: `ThreadLocalRandom` shouldn't be used that way. Please read documentation - current should be invoked for each invocation of next()

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+  override def processStateDefinitionManager: ProcessStateDefinitionManager = SimpleProcessStateDefinitionManager
+
+  override def customActions: List[CustomAction] = List(
+    CustomAction("test-running", List("RUNNING")),
+    CustomAction("test-canceled", List("CANCELED")),
+    CustomAction("test-all", Nil),
+  )
+
+  override def invokeCustomAction(actionRequest: CustomActionRequest, canonicalProcess: CanonicalProcess): Future[Either[CustomActionError, CustomActionResult]] =
```
Comment (arkadius) [inline engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:86]: Can't you do `asyncChangeState` in this action invocations? (based on `canonicalProcess.id`)? I thought that it is the main purpose of this class - to be able to see that custom action "do sth". How u will know that something was happen?

**Comment by arkadius** [comment]
Can u change the title of this PR to "Development Deployment Manager"? Engine looks like it executes scenario and it just fake deployment

File: engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala (hunk 4 above)
Comment (lciolecki) [reply engine/development/deploymentManager/src/main/scala/pl/touk/nussknacker/development/manager/DevelopmentDeploymentManagerProvider.scala:31]: Okey, thanks.

**Review by arkadius — APPROVED** [review]

**Pull request merged**